### PR TITLE
[SPARK-9193] Avoid assigning tasks to "lost" executor(s)

### PR DESCRIPTION
**Pull request by GraceH**
Now, when some executors are killed by dynamic-allocation, it leads to some mis-assignment onto lost executors sometimes. Such kind of mis-assignment causes task failure(s) or even job failure if it repeats that errors for 4 times. 

The root cause is that **_killExecutors**_ doesn't remove those executors under killing ASAP. It depends on the **_OnDisassociated**_ event to refresh the active working list later. The delay time really depends on your cluster status (from several milliseconds to sub-minute). When new tasks to be scheduled during that period of time, it will be assigned to those "active" but "under killing" executors. Then the tasks will be failed due to "executor lost". The better way is to exclude those executors under killing in the makeOffers(). Then all those tasks won't be allocated onto those executors "to be lost" any more.

### CODE REVIEW

**Comment by squito** [comment]
Hi @GraceH thanks for finding this bug and posting the PR.  The changes look completely reasonable, but would it be possible to add a unit test?  I know this is a big ask, since it doesn't look like there are any tests right now on that test anything like this for `CoarseGrainedSchedulerBackend`, so it will probably take some work to do.

I believe this changes are correct, so the test isn't so much about verifying this fix as it is to prevent future regressions.  It would also really help in the future for following along the code -- eg., it took me a while to figure out how anything was ever removed from `executorsPendingToRemove`.

**Comment by GraceH** [comment]
@squito Thanks for the confirmation. If you do think it is necessary to add unit test for ```CoarseGrainedSchedulerBackend```, one more JIRA can be filed for that. Because it seems that should cover more than what this PR does.  What do you think?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 1]
```diff
@@ -169,8 +169,11 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
 
     // Make fake resource offers on all executors
     private def makeOffers() {
-      launchTasks(scheduler.resourceOffers(executorDataMap.map { case (id, executorData) =>
-        new WorkerOffer(id, executorData.executorHost, executorData.freeCores)
+      launchTasks(scheduler.resourceOffers(executorDataMap
+        // Filter out executors under killing
+        .filterKeys(!executorsPendingToRemove.contains(_))
+        .map { case (id, executorData) =>
+          new WorkerOffer(id, executorData.executorHost, executorData.freeCores)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: This is a little hard to read. Could you rewrite this as follows:
```
private def makeOffers() {
  val activeExecutors = executorDataMap.filterKeys(!executorsPendingToRemove)
  val workerOffers = activeExecutors.map { case (id, executorData) =>
    new WorkerOffer(id, ...)
  }
  launchTasks(scheduler.resourceOffers(workerOffers))
}
```

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 2]
```diff
@@ -181,9 +184,12 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
...
     // Make fake resource offers on just one executor
     private def makeOffers(executorId: String) {
-      val executorData = executorDataMap(executorId)
-      launchTasks(scheduler.resourceOffers(
-        Seq(new WorkerOffer(executorId, executorData.executorHost, executorData.freeCores))))
+      // Filter out executors under killing
+      if (!executorsPendingToRemove.contains(executorId)) {
+        val executorData = executorDataMap(executorId)
+        launchTasks(scheduler.resourceOffers(
+          Seq(new WorkerOffer(executorId, executorData.executorHost, executorData.freeCores))))
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: same here:
```
if (!executorsPendingToRemove.contains(executorId)) {
  val executorData = executorDataMap(executorId)
  val workerOffers = Seq(new WorkerOffer(executorId, ...)
  launchTasks(scheduler.resourceOffers(workerOffers))
}
```

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala (hunk 1 above)
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: Good point. Will re-organize the code snippet.

**Comment by andrewor14** [comment]
Hi @GraceH this looks great. I left a few minor wording suggestions to improve readability.

As for unit tests, I do agree that we should have them, but it seems outside the scope of this fix since properly testing all of this logic would require a non-trivial refactor. In this case, I am more inclined to merge in this fix than to delay it indefinitely.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala (hunk 2 above)
Comment (GraceH) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: Though this one is kept as before, I will re-format that in this PR.

**Comment by GraceH** [comment]
Thanks @andrewor14 . That makes sense. Will have a revised version with more readable lines.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala (hunk 2 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: yup, the existing code was already difficult to read.

**Comment by squito** [comment]
yeah, I don't love the idea of adding things w/out tests, but in this case I suppose its best left for the future.  lgtm pending the tests passing

**Comment by GraceH** [comment]
Thanks @squito.

**Pull request closed without merging**